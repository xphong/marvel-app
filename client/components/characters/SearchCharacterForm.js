import React, { Component } from 'react';
import ReactDOM from 'react-dom';
import { browserHistory } from 'react-router';

export default class SearchCharacterForm extends Component {
  constructor() {
    super();

    this.handleSearch = this.handleSearch.bind(this);
    this.handleInputChange = this.handleInputChange.bind(this);
  }

  componentDidMount() {
    this.setState({
      characterName: null
    });
  }

  handleInputChange(event) {
    this.setState({
      [event.target.name]: event.target.value
    });
  }

  handleSearch(event) {
    event.preventDefault();

    this.props.actions.fetchCharactersByName(this.state.characterName);
    browserHistory.push(`/characters/${this.state.characterName}`);

    this.resetForm();
  }

  resetForm() {
    ReactDOM.findDOMNode(this.refs.searchCharacterForm).reset();
    ReactDOM.findDOMNode(this.refs.characterName).focus();
    this.state = { characterName: null };
  }

  render() {
    return (
      <div className="search-character__form">
        <form ref="searchCharacterForm"
          onSubmit={this.handleSearch}
          onChange={this.handleInputChange}
        >

          <div className="ui action input">
            <input ref="characterName"
              name="characterName"
              placeholder="Character Name"
              type="text"
              required
            />

              <button className="ui icon pink button">
                  <i className="search icon" />
              </button>
          </div>
        </form>
      </div>
    );
  }
}
