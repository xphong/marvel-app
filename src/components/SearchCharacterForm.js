import React, { Component } from 'react';
import ReactDOM from 'react-dom';

export default class SearchCharacterForm extends Component {
  constructor(props, context) {
    super(props, context);
    this.state = {characterName: null};
  }

  handleInputChange(event) {
    this.setState({
      [event.target.name]: event.target.value
    });
  }

  handleSearch(event) {
    event.preventDefault();

    this.props.actions.fetchCharactersByName(this.state.characterName);
    this.resetForm();
  }

  resetForm() {
    ReactDOM.findDOMNode(this.refs.searchCharacterForm).reset();
    ReactDOM.findDOMNode(this.refs.characterName).focus();
    this.state = {characterName: null};
  }

  render() {
    return (
      <div className='search-character__form'>
        <form ref='searchCharacterForm'
              onSubmit={this.handleSearch.bind(this)}
              onChange={this.handleInputChange.bind(this)}>

          <div className='ui action input'>
            <input ref='characterName'
                   name='characterName'
                   placeholder='Character Name'
                   type='text'
                   required
                   />

              <button className='ui icon pink button'>
                  <i className='search icon'></i>
              </button>
          </div>
        </form>
      </div>
    );
  }
}
