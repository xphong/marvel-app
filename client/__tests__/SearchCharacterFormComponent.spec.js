import React from 'react';
import { browserHistory } from 'react-router';

import SearchCharacterForm from '../components/characters/SearchCharacterForm';

describe('SearchCharacterForm Component', () => {
  it('should render SearchCharacterForm', () => {
    const wrapper = shallow(
      <SearchCharacterForm />
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should perform a search', () => {
    const wrapper = shallow(
      <SearchCharacterForm />
    );

    const value = 'Professor X';

    wrapper.find('input').simulate('change', {
      target: {
        name: 'characterName',
        value
      },
    });

    wrapper.find('button').simulate('click');

    expect(wrapper).toMatchSnapshot();
  });

  describe('handleInputChange', () => {
    it('should update characterName state', () => {
      const wrapper = shallow(
        <SearchCharacterForm />
      );

      const value = 'Magneto';

      wrapper.instance().handleInputChange({
        target: {
          name: 'characterName',
          value
        }
      });

      expect(wrapper.state().characterName).toEqual(value);
    });
  });

  describe('handleSearch', () => {
    function handleSearchSetup() {
      const props = {
        actions: {
          fetchCharactersByName: jest.fn()
        }
      };

      const wrapper = shallow(
        <SearchCharacterForm {...props} />
      );

      spyOn(browserHistory, 'push');
      spyOn(wrapper.instance(), 'resetForm');

      wrapper.state().characterName = 'Rogue';
      wrapper.instance().handleSearch({
        preventDefault: jest.fn()
      });

      return {
        props,
        wrapper
      };
    }

    it('should call fetchCharactersByName action', () => {
      const { wrapper, props } = handleSearchSetup();
      expect(props.actions.fetchCharactersByName).toHaveBeenCalledWith(wrapper.state().characterName);
    });

    it('should call browserHistory', () => {
      const { wrapper } = handleSearchSetup();
      expect(browserHistory.push).toHaveBeenCalledWith(`/characters/${wrapper.state().characterName}`);
    });

    it('should call resetForm', () => {
      const { wrapper } = handleSearchSetup();
      expect(wrapper.instance().resetForm).toHaveBeenCalled();
    });
  });
});
