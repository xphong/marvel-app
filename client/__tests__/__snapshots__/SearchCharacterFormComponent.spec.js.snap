// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchCharacterForm Component should perform a search 1`] = `
<div
  className="search-character__form"
>
  <form
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="ui action input"
    >
      <input
        name="characterName"
        placeholder="Character Name"
        required={true}
        type="text"
      />
      <button
        className="ui icon pink button"
      >
        <i
          className="search icon"
        />
      </button>
    </div>
  </form>
</div>
`;

exports[`SearchCharacterForm Component should render SearchCharacterForm 1`] = `
<div
  className="search-character__form"
>
  <form
    onChange={[Function]}
    onSubmit={[Function]}
  >
    <div
      className="ui action input"
    >
      <input
        name="characterName"
        placeholder="Character Name"
        required={true}
        type="text"
      />
      <button
        className="ui icon pink button"
      >
        <i
          className="search icon"
        />
      </button>
    </div>
  </form>
</div>
`;
